### PR TITLE
Closes #923 Subject ID format

### LINKED ISSUE #923: BIDS2Legacy issue with certain subject names
### Description

BIDS2Legacy crashes for `Philips_PCASL_2DEPI_3.2.2.1_2` because it doesn't convert anything from BIDS.

- The issue is the name 'EMI_301' - the '_' is lost during the process...
- Dcm2niix still creates 'EMI_301' in `temp`.
- In rawdata, I already have 'sub_EMI301' and not 'sub_EMI_301' so later, BIDS2Legacy relies on information from sourcedata (wrongly) and doesn't find the subject in rawdata. 
- We need to skip the `_` and replace by an empty character in these places:
   - reading subject structure
   - DCM2NII
   - NII2BIDS
- Function `xASL_imp_CheckForAliasInVisit` inside `xASL_imp_NII2BIDS_Subject` seems to be doing that for me.
- At line 92, subjectName is still `EMI_301`, but `bidsLabel.subject = xASL_adm_CorrectName` shortens it to `EMI301`.

What I propose is that we need to remove all `_` in patient names systematically and also in session/visit names as these are also in BIDS.

### Release notes

Improve import modularity and BIDS format of subject IDs.

### PR DESCRIPTION
**Pull request by MichaelStritt**
### Linked issue

Check out #923

### How to test

Run the flavors import.

### CODE REVIEW

**Comment by MichaelStritt** [comment]
Okay, so my idea would be the following:

- If we want to keep `sourcedata` as `read-only` and we want to allow subject names including `-` or `_` or something else, then we should basically keep DCM2NII as it is right now. The reading of the `sourcedata` does not have to change, but maybe the folder structure that we output, because we need to work on NII2BIDS anyway, so that it can read the `tempdata` without a `sourceStructure.json` or anything else. DCM2NII could basically already store the NIFTIs and JSONS as sub/ses/run, if we know that at that point already. This would mean that we need to rewrite some parts of `xASL_imp_DetermineStructureFromTempdata` and we would get rid of the underscores and so on in this intermediate step before we get to BIDS.
- Alternatively we keep all of the mentioned parts of the workflow and instead we just get rid of the special characters during NII2BIDS. The modularity of NII2BIDS still wouldn't be there though. Kind of a tricky thing.

**Comment by MichaelStritt** [comment]
@jan-petr: Please have a look at d4e689b. It is supposed to help with modularity & getting the subject id to the bids format. DCM2NII depends on the `sourceStructure.json`, but NII2BIDS, Deface and BIDS2Legacy have to be independent. This should work with this commit. In addition, the subject IDs are made BIDS compliant within DCM2NII now. This is done using an "export id" for the directory name. Think the code is quite clean and helps with comments from [here](https://github.com/ExploreASL/ExploreASL/pull/943#issuecomment-979080276) and is somewhat related to [this](https://github.com/ExploreASL/ExploreASL/pull/937#issuecomment-979454722).

**Comment by MichaelStritt** [comment]
@jan-petr: This is kind of related to #945, so it would be nice to merge this (maybe today?), so that I can get some work done for #945. I'm testing right now if #923 doesn't break other flavors.

File: Modules/SubModule_Import/xASL_imp_DCM2NII_Subject.m [hunk 1]
```diff
@@ -146,17 +149,33 @@
...
+
+    % If a subject ID has a `_` or a `-` or something else, we need to get
+    % rid of it in the directory name for BIDS, for the import of the
+    % sourcedata we should still know about that though, which is why we
+    % use a new variable for the subject ID directory export.
+    
+    subjectExport = xASL_adm_CorrectName(subjectID, 2);
+    
+    if ~strcmp(subjectID,subjectExport)
+        fprintf(2,'Special characters in subject ID...\n');
```
Comment (jan-petr) [inline Modules/SubModule_Import/xASL_imp_DCM2NII_Subject.m:175]: It is really helpful for the user to print something like:
Special characters in subject ID, changing 'subject_name' -> 'subjectname'

**Review by jan-petr — CHANGES_REQUESTED** [review]
Minor reporting comment.

File: Modules/SubModule_Import/xASL_imp_DCM2NII_Subject.m (hunk 1 above)
Comment (MichaelStritt) [reply Modules/SubModule_Import/xASL_imp_DCM2NII_Subject.m:175]: Good idea, will do 👍

**Comment by MichaelStritt** [comment]
@jan-petr: Okay, so if we ignore the PAR REC datasets and the Hadamards as well, then there's only a few left.

```
COMPARISON TABLE:
                        flavor                                dataset                      name                                            message                              
    _______________________________________________    ______________________    ________________________    ___________________________________________________________________

    'Philips_PCASL_2DEPI_3.2.2.1_2'                    'Both'                    'Different file content'    'Different file content: dataset_description.json'                 
    'Philips_PCASL_2DEPI_PARREC_1'                     'rawdata'                 'Missing folder'            '\sub-Sub01\ses-1\anat'                                            
    'Philips_PCASL_2DEPI_PARREC_1'                     'rawdata'                 'Missing file'              '\sub-Sub01\ses-1\anat\sub-Sub01_ses-1_T1w.json'                   
    'Philips_PCASL_2DEPI_PARREC_1'                     'rawdata'                 'Missing file'              '\sub-Sub01\ses-1\anat\sub-Sub01_ses-1_T1w.nii.gz'                 
    'Philips_PCASL_2DEPI_PARREC_1'                     'rawdata'                 'Missing file'              '\sub-Sub01\ses-1\perf\sub-Sub01_ses-1_asl.json'                   
    'Philips_PCASL_2DEPI_PARREC_1'                     'rawdata'                 'Missing file'              '\sub-Sub01\ses-1\perf\sub-Sub01_ses-1_asl.nii.gz'                 
    'Philips_PCASL_2DEPI_PARREC_1'                     'rawdata'                 'Missing file'              '\sub-Sub01\ses-1\perf\sub-Sub01_ses-1_aslcontext.tsv'             
    'Philips_PCASL_2DEPI_PARREC_2'                     'rawdata'                 'Missing file'              '\sub-Sub01\perf\sub-Sub01_asl.json'                               
    'Philips_PCASL_2DEPI_PARREC_2'                     'rawdata'                 'Missing file'              '\sub-Sub01\perf\sub-Sub01_asl.nii.gz'                             
    'Philips_PCASL_2DEPI_PARREC_2'                     'rawdata'                 'Missing file'              '\sub-Sub01\perf\sub-Sub01_aslcontext.tsv'                         
    'Siemens_PCASL_3DGRASE_VD13A_Hadamard4-FME_1'      'rawdata'                 'Missing file'              '\sub-Sub1\perf\sub-Sub1_asl.json'                                 
    'Siemens_PCASL_3DGRASE_VD13A_Hadamard4-FME_1'      'rawdata'                 'Missing file'              '\sub-Sub1\perf\sub-Sub1_asl.nii.gz'                               
    'Siemens_PCASL_3DGRASE_VD13A_Hadamard4-FME_1'      'rawdata'                 'Missing file'              '\sub-Sub1\perf\sub-Sub1_aslcontext.tsv'                           
    'Siemens_PCASL_3DGRASE_VD13A_Hadamard4-FME_1'      'rawdata'                 'Missing file'              '\sub-Sub1\perf\sub-Sub1_m0scan.json'                              
    'Siemens_PCASL_3DGRASE_VD13A_Hadamard4-FME_1'      'rawdata'                 'Missing file'              '\sub-Sub1\perf\sub-Sub1_m0scan.nii.gz'                            
    'Siemens_PCASL_3DGRASE_VD13A_Hadamard8-FME_1'      'rawdata'                 'Missing file'              '\sub-Sub1\perf\sub-Sub1_asl.json'                                 
    'Siemens_PCASL_3DGRASE_VD13A_Hadamard8-FME_1'      'rawdata'                 'Missing file'              '\sub-Sub1\perf\sub-Sub1_asl.nii.gz'                               
    'Siemens_PCASL_3DGRASE_VD13A_Hadamard8-FME_1'      'rawdata'                 'Missing file'              '\sub-Sub1\perf\sub-Sub1_aslcontext.tsv'                           
    'Siemens_PCASL_3DGRASE_VD13A_Hadamard8-FME_1'      'rawdata'                 'Missing file'              '\sub-Sub1\perf\sub-Sub1_m0scan.json'                              
    'Siemens_PCASL_3DGRASE_VD13A_Hadamard8-FME_1'      'rawdata'                 'Missing file'              '\sub-Sub1\perf\sub-Sub1_m0scan.nii.gz'                            
    'GE_PCASL_3Dspiral_24.0LX_WIP_1'                   'derivativesReference'    'Missing folder'            '\ExploreASL\temp'                                                 
    'GE_PCASL_3Dspiral_25.0LX_ADNI3_027_S_5079_1'      'derivativesReference'    'Missing file'              '\ExploreASL\sub-001\ASL_1\ASL4D.nii.gz'                           
    'GE_PCASL_3Dspiral_25.0LX_ADNI3_027_S_5079_1'      'derivativesReference'    'Missing file'              '\ExploreASL\sub-001\ASL_1\ASL4D_aslcontext.tsv'                   
    'GE_PCASL_3Dspiral_25.0LX_ADNI3_027_S_5079_1'      'derivatives'             'Missing file'              '\ExploreASL\sub-001\ASL_1\ASL4D.nii'                              
    'GE_PCASL_3Dspiral_25.0LX_ADNI3_027_S_5079_1'      'derivatives'             'Missing file'              '\ExploreASL\sub-001\ASL_1\ASL4D_Source.json'                      
    'GE_PCASL_3Dspiral_25.0LX_ADNI3_027_S_5079_1'      'derivatives'             'Missing file'              '\ExploreASL\sub-001\ASL_1\ASL4D_Source.nii'                       
    'GE_PCASL_3Dspiral_25.0LX_ADNI3_027_S_5079_1'      'derivatives'             'Missing file'              '\ExploreASL\sub-001\ASL_1\ASL4D_Source_aslcontext.tsv'            
    'GE_PCASL_3Dspiral_25.0LX_ADNI3_027_S_5079_1'      'derivatives'             'Missing file'              '\ExploreASL\sub-001\ASL_1\M0.json'                                
    'GE_PCASL_3Dspiral_25.0LX_ADNI3_027_S_5079_1'      'derivatives'             'Missing file'              '\ExploreASL\sub-001\ASL_1\M0.nii'                                 
    'GE_PCASL_3Dspiral_25.0LX_ADNI3_027_S_5079_1'      'Both'                    'Different file content'    'Different file content: \ExploreASL\sub-001\ASL_1\ASL4D.json'     
    'Philips_PASL_2DEPI_5.4.1.0_ADNI3_006_S_6681_1'    'Both'                    'Different file content'    'Different file content: \ExploreASL\sub-001\ASL_1\ASL4D.json'     
    'Philips_PCASL_2DEPI_3.2.2.1_2'                    'derivativesReference'    'Missing folder'            '\ExploreASL\temp'                                                 
    'Philips_PCASL_2DEPI_3.2.2.1_2'                    'Both'                    'Different file content'    'Different file content: \ExploreASL\dataset_description.json'     
    'Philips_PCASL_2DEPI_3.2.2.1_2'                    'Both'                    'Different file content'    'Different file content: \ExploreASL\sub-EMI301_1\ASL_1\ASL4D.json'
    'Philips_PCASL_2DEPI_3.2.2.1_2'                    'Both'                    'Different file content'    'Different file content: \ExploreASL\sub-EMI301_1\ASL_1\M0.json'   
    'Philips_PCASL_2DEPI_3.2.2.1_2'                    'Both'                    'Different file content'    'Different file content: \ExploreASL\sub-EMI301_1\FLAIR.json'      
    'Philips_PCASL_2DEPI_3.2.2.1_2'                    'Both'                    'Different file content'    'Different file content: \ExploreASL\sub-EMI301_1\T1.json'         
    'Philips_PCASL_2DEPI_5.1.9.1_GMB_1'                'derivativesReference'    'Missing folder'            '\ExploreASL\temp'                                                 
    'Philips_PCASL_2DEPI_PARREC_1'                     'derivativesReference'    'Missing folder'            '\ExploreASL\temp'                                                 
    'Philips_PCASL_2DEPI_PARREC_1'                     'derivatives'             'Missing folder'            '\ExploreASL\sub-Sub01\ASL_1'                                      
    'Philips_PCASL_2DEPI_PARREC_1'                     'derivativesReference'    'Missing file'              '\ExploreASL\temp\dcm2niiCatchedErrors.json'                       
    'Philips_PCASL_2DEPI_PARREC_1'                     'derivativesReference'    'Missing file'              '\ExploreASL\temp\dcm2niiCatchedErrors.mat'                        
    'Philips_PCASL_2DEPI_PARREC_1'                     'derivatives'             'Missing file'              '\ExploreASL\sub-Sub01\ASL_1\ASL4D.json'                           
    'Philips_PCASL_2DEPI_PARREC_1'                     'derivatives'             'Missing file'              '\ExploreASL\sub-Sub01\ASL_1\ASL4D.nii.gz'                         
    'Philips_PCASL_2DEPI_PARREC_1'                     'derivatives'             'Missing file'              '\ExploreASL\sub-Sub01\ASL_1\ASL4D_aslcontext.tsv'                 
    'Philips_PCASL_2DEPI_PARREC_1'                     'derivatives'             'Missing file'              '\ExploreASL\sub-Sub01\T1.json'                                    
    'Philips_PCASL_2DEPI_PARREC_1'                     'derivatives'             'Missing file'              '\ExploreASL\sub-Sub01\T1.nii.gz'                                  
    'Philips_PCASL_2DEPI_PARREC_2'                     'derivativesReference'    'Missing folder'            '\ExploreASL\temp'                                                 
    'Philips_PCASL_2DEPI_PARREC_2'                     'derivatives'             'Missing folder'            '\ExploreASL\sub-Sub01_1\ASL_1'                                    
    'Philips_PCASL_2DEPI_PARREC_2'                     'derivativesReference'    'Missing file'              '\ExploreASL\temp\dcm2niiCatchedErrors.json'                       
    'Philips_PCASL_2DEPI_PARREC_2'                     'derivativesReference'    'Missing file'              '\ExploreASL\temp\dcm2niiCatchedErrors.mat'                        
    'Philips_PCASL_2DEPI_PARREC_2'                     'derivatives'             'Missing file'              '\ExploreASL\sub-Sub01_1\ASL_1\ASL4D.json'                         
    'Philips_PCASL_2DEPI_PARREC_2'                     'derivatives'             'Missing file'              '\ExploreASL\sub-Sub01_1\ASL_1\ASL4D.nii.gz'                       
    'Philips_PCASL_2DEPI_PARREC_2'                     'derivatives'             'Missing file'              '\ExploreASL\sub-Sub01_1\ASL_1\ASL4D_aslcontext.tsv'               
    'Philips_PCASL_3DGRASE_5.4.0.2_1'                  'derivativesReference'    'Missing folder'            '\ExploreASL\temp'                                                 
    'Siemens_PASL_2DEPI_VB15A_ADNI2_941_S_4764_1'      'Both'                    'Different file content'    'Different file content: \ExploreASL\dataPar.json'                 
    'Siemens_PCASL_3DGRASE_VD13A_Hadamard4-FME_1'      'derivativesReference'    'Missing folder'            '\ExploreASL\temp'                                                 
    'Siemens_PCASL_3DGRASE_VD13A_Hadamard4-FME_1'      'derivatives'             'Missing folder'            '\ExploreASL\sub-Sub1_1\ASL_1'                                     
    'Siemens_PCASL_3DGRASE_VD13A_Hadamard4-FME_1'      'derivativesReference'    'Missing file'              '\ExploreASL\temp\dcm2niiCatchedErrors.json'                       
    'Siemens_PCASL_3DGRASE_VD13A_Hadamard4-FME_1'      'derivativesReference'    'Missing file'              '\ExploreASL\temp\dcm2niiCatchedErrors.mat'                        
    'Siemens_PCASL_3DGRASE_VD13A_Hadamard4-FME_1'      'derivatives'             'Missing file'              '\ExploreASL\sub-Sub1_1\ASL_1\ASL4D.json'                          
    'Siemens_PCASL_3DGRASE_VD13A_Hadamard4-FME_1'      'derivatives'             'Missing file'              '\ExploreASL\sub-Sub1_1\ASL_1\ASL4D.nii.gz'                        
    'Siemens_PCASL_3DGRASE_VD13A_Hadamard4-FME_1'      'derivatives'             'Missing file'              '\ExploreASL\sub-Sub1_1\ASL_1\ASL4D_aslcontext.tsv'                
    'Siemens_PCASL_3DGRASE_VD13A_Hadamard4-FME_1'      'derivatives'             'Missing file'              '\ExploreASL\sub-Sub1_1\ASL_1\M0.json'                             
    'Siemens_PCASL_3DGRASE_VD13A_Hadamard4-FME_1'      'derivatives'             'Missing file'              '\ExploreASL\sub-Sub1_1\ASL_1\M0.nii.gz'                           
    'Siemens_PCASL_3DGRASE_VD13A_Hadamard8-FME_1'      'derivativesReference'    'Missing folder'            '\ExploreASL\temp'                                                 
    'Siemens_PCASL_3DGRASE_VD13A_Hadamard8-FME_1'      'derivatives'             'Missing folder'            '\ExploreASL\sub-Sub1_1\ASL_1'                                     
    'Siemens_PCASL_3DGRASE_VD13A_Hadamard8-FME_1'      'derivativesReference'    'Missing file'              '\ExploreASL\temp\dcm2niiCatchedErrors.json'                       
    'Siemens_PCASL_3DGRASE_VD13A_Hadamard8-FME_1'      'derivativesReference'    'Missing file'              '\ExploreASL\temp\dcm2niiCatchedErrors.mat'                        
    'Siemens_PCASL_3DGRASE_VD13A_Hadamard8-FME_1'      'derivatives'             'Missing file'              '\ExploreASL\sub-Sub1_1\ASL_1\ASL4D.json'                          
    'Siemens_PCASL_3DGRASE_VD13A_Hadamard8-FME_1'      'derivatives'             'Missing file'              '\ExploreASL\sub-Sub1_1\ASL_1\ASL4D.nii.gz'                        
    'Siemens_PCASL_3DGRASE_VD13A_Hadamard8-FME_1'      'derivatives'             'Missing file'              '\ExploreASL\sub-Sub1_1\ASL_1\ASL4D_aslcontext.tsv'                
    'Siemens_PCASL_3DGRASE_VD13A_Hadamard8-FME_1'      'derivatives'             'Missing file'              '\ExploreASL\sub-Sub1_1\ASL_1\M0.json'                             
    'Siemens_PCASL_3DGRASE_VD13A_Hadamard8-FME_1'      'derivatives'             'Missing file'              '\ExploreASL\sub-Sub1_1\ASL_1\M0.nii.gz'                           
    'Siemens_PCASL_3DGRASE_VE11C_1'                    'derivativesReference'    'Missing folder'            '\ExploreASL\temp'                                                 

LOGGING TABLE:
```

- `Siemens_PCASL_3DGRASE_VE11C_1`: Seems like it just didn't fully delete temp. The directory is there but it is empty.
- `Philips_PCASL_3DGRASE_5.4.0.2_`: Same here.
- `Philips_PCASL_2DEPI_5.1.9.1_GMB_1`: Same here.
- `GE_PCASL_3Dspiral_24.0LX_WIP_1`: Same here.
- `Philips_PCASL_2DEPI_3.2.2.1_2`: Seems like here there's just the version information of ExploreASL in the JSONs and that one differs.
- `GE_PCASL_3Dspiral_25.0LX_ADNI3_027_S_5079_1`: Weirdly the ASL4D source NIFTI isn't there and the M0 is missing too, did we change something there?

**Review by jan-petr — APPROVED** [review]
OK.

**Pull request merged**